rfd/0085-improve-webapps-codebase.md

```diff
@@ -0,0 +1,163 @@
+    fetchAttempt={{ status: 'failed', statusText: 'some fetch  error' }}
+    tdpConnection={{ status: 'success' }}
+    wsConnection={'open'}
+    disconnected={false}
+  />
+);
+```
+
+Which is then tested like so:
+
+```typescript jsx
+test('fetch error', () => {
+  const { getByTestId } = render(<FetchError />);
+  expect(getByTestId('Modal')).toMatchSnapshot();
+});
+```
+
+This results in all the logic in `useDesktopSession` going untested. We're only testing the UI part of the application,
+but skipping all of the business logic.
+
+Storybook isn't best suited for the way we are using it. As a result of testing our large components primarily via stories, we have
+ended up only testing our UI and nothing else.
+
+## Details
+
+### Changing how we write stories
+
+Experience tells us that Stories in Storybook are used most effectively for documenting how to use reusable components. We should therefore restrict usage of Storybook mostly to
+our shared component library, and test our fullscreen components another way.
+
+### Changing how we write tests
+
+We should be testing all the different state possibilities in Jest and React Testing Library, mocking any data or 
+network requests to cause the component to render into the state we want. This ensures that the internal logic
+of our components (generally the most critical and difficult-to-get-right aspect of frontend code) gets tested,
+as well as the final UI state.
+
+Tests should check for elements that exist for the specific state being tested, i.e. an error message when the data has 
+failed to load. Elements should be clicked on or events triggered in order to change the state of the component.
+
+By writing tests as described above, snapshot tests typically become redundant and therefore unnecessary. Besides the problem of skipping the business logic, snapshots consistently break for non-bugs, such as when styles are updated, and
+generally don't provide much if any security around behaviour or appearance. Further reading - 
+https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e.
+
+### Changing how we export components
+
+Default exports should be avoided except when absolutely necessary (for example, with 
+`React.Lazy`), as they degrade the developer experience
+for no discernible benefit. For more fully fleshed out reasoning, refer to the articles below:
+- https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
+- https://rajeshnaroth.medium.com/avoid-es6-default-exports-a24142978a7a
+- https://blog.piotrnalepa.pl/2020/06/26/default-exports-vs-named-exports/
+- https://ilikekillnerds.com/2019/08/default-exports-bad/
+
+When needing to export something as default, we should export it like so:
+
+```typescript jsx
+function SomeComponent() {
+  return (
+    <div>
+      hello!
+    </div>
+  );
+}
+
+export { SomeComponent as default };
+```
+
+This provides a clearer, more explicit export.
+
+### Code Coverage
```

Comment:
I'm not a fan of code coverage as it is a vanity metric. Having 100% coverage doesn't mean that you have good tests. I'd prefer we focus on ensuring we have tested the functionalities we expect than try to optimize for coverage.

Reply:
There's a balance between the effort to write the tests vs the coverage that you get. I wouldn't ever want to enforce 100% - you'd write tests just for the sake of it. 

Enforcing say, 70%, would ensure that decent enough tests are being written that hit enough of the branches we truly care about, and also ensures that we aren't skipping writing tests as we've not had any coverage on our custom hooks _at all_ without knowing about it. 

Also, code coverage at least gives us an insight into where we are completely lacking coverage - it's a really useful tool, and if configured correctly can be an added benefit to the codebase and not a cumbersome PITA.

Also nothing is set in stone, so if it's causing issues or people are writing tests just to please, we can move the barrier, or reassess if it's worth enforcing at all. But at least having an interactive UI and an insight into our coverage stats is a very useful metric.

Reply:
+1 on displaying it, strongly against enforcing it. Enforcement of some arbitrary percentage makes it difficult for us to make judgement calls about what to test (@hatched's point) and also when to test (i.e. if we want to prototype a complex feature quickly and then write tests later).

Reply:
I've removed the section about enforcement